echidna/limit/minimise.py

```diff
@@ -0,0 +1,317 @@
+""" Module containing classes that act as minimisers in a fit.
+"""
+import numpy
+
+from echidna.limit.fit_results import FitResults
+
+import copy
+import abc
+
+
+class Minimiser(object):
+    """ Base class for minimiser objects.
+
+    Args:
+      name (string): Name of minimiser.
+      per_bin (bool, optional): Flag if minimiser should expect a
+        test statistic value per-bin.
+
+    Attributes:
+      _name (string): Name of minimiser.
+      _per_bin (bool, optional): Flag if minimiser should expect a
+        test statistic value per-bin.
+      _type (string): Type of minimiser, e.g. GridSearch
+    """
+    __metaclass__ = abc.ABCMeta  # Only required for python 2
+
+    def __init__(self, name, per_bin=False):
+        self._name = name
+        self._type = None  # No type for base class
+        self._per_bin = per_bin
+
+    @abc.abstractmethod
+    def minimise(self, funct):
```

Comment:
Should add a test_statistic argument to keep consistent with overriding classes.